In [41]:

!pip install tensorflow-gpu==2.0.0-rc1

In [0]:
from tensorflow.keras.models import Sequential 

In [0]:

import warnings
warnings.filterwarnings("ignore")

from tensorflow.keras.datasets import mnist

In [0]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

input_dim = 784  # 28*28
output_dim = nb_classes = 10
nb_epoch = 20

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

Assignments
In this assignment, you'll continue working on the MNIST dataset. So, train and test the models specified in the tasks below using the MNIST data. Although you can choose the number of epochs for training, we recommend training 20 epochs for the sake of comparison.

1. Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using 8 as the mini batch size.

In [0]:
from tensorflow.keras.utils import to_categorical

Y_train = to_categorical(y_train, nb_classes)
Y_test = to_categorical(y_test, nb_classes)

In [46]:
X_train[0].shape

(784,)

In [0]:
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])


In [48]:
# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=8, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 14s 234us/sample - loss: 0.3645 - accuracy: 0.8974
Epoch 2/20
60000/60000 [==============================] - 14s 231us/sample - loss: 0.1700 - accuracy: 0.9504
Epoch 3/20
60000/60000 [==============================] - 14s 237us/sample - loss: 0.1223 - accuracy: 0.9645
Epoch 4/20
60000/60000 [==============================] - 14s 232us/sample - loss: 0.0952 - accuracy: 0.9726
Epoch 5/20
60000/60000 [==============================] - 14s 230us/sample - loss: 0.0776 - accuracy: 0.9778
Epoch 6/20
60000/60000 [==============================] - 14s 228us/sample - loss: 0.0649 - accuracy: 0.9809
Epoch 7/20
60000/60000 [==============================] - 14s 226us/sample - loss: 0.0549 - accuracy: 0.9841
Epoch 8/20
60000/60000 [==============================] - 14s 226us/sample - loss: 0.0478 - accuracy: 0.9863
Epoch 9/20
60000/60000 [==============================] - 14s 227us/sample - loss: 0.0413 - accuracy: 0.9

In [49]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.07742905021063926
Test accuracy: 0.9784


2. Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using 128 as the mini batch size.

In [50]:
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 33us/sample - loss: 1.2680 - accuracy: 0.6747
Epoch 2/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.5005 - accuracy: 0.8702
Epoch 3/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.3875 - accuracy: 0.8933
Epoch 4/20
60000/60000 [==============================] - 2s 30us/sample - loss: 0.3406 - accuracy: 0.9048
Epoch 5/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.3119 - accuracy: 0.9117
Epoch 6/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.2914 - accuracy: 0.9175
Epoch 7/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.2748 - accuracy: 0.9218
Epoch 8/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.2608 - accuracy: 0.9262
Epoch 9/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.2485 - accuracy: 0.9297
Epoch 10/20
60

In [51]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.16713767684400083
Test accuracy: 0.9511


3. Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using the full sample as the batch.

In [52]:
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, X_train.shape[0], epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 19us/sample - loss: 2.3406 - accuracy: 0.1538
Epoch 2/20
60000/60000 [==============================] - 1s 12us/sample - loss: 2.3336 - accuracy: 0.1568
Epoch 3/20
60000/60000 [==============================] - 1s 12us/sample - loss: 2.3268 - accuracy: 0.1596
Epoch 4/20
60000/60000 [==============================] - 1s 11us/sample - loss: 2.3201 - accuracy: 0.1623
Epoch 5/20
60000/60000 [==============================] - 1s 12us/sample - loss: 2.3136 - accuracy: 0.1652
Epoch 6/20
60000/60000 [==============================] - 1s 12us/sample - loss: 2.3072 - accuracy: 0.1683
Epoch 7/20
60000/60000 [==============================] - 1s 11us/sample - loss: 2.3010 - accuracy: 0.1711
Epoch 8/20
60000/60000 [==============================] - 1s 12us/sample - loss: 2.2949 - accuracy: 0.1743
Epoch 9/20
60000/60000 [==============================] - 1s 11us/sample - loss: 2.2889 - accuracy: 0.1776
Epoch 10/20
60

In [53]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 2.2233628135681154
Test accuracy: 0.212


RESULTS: Using batchsize=8, accuracy =0.978
      Using batchsize=128, accuracy =0.951
      Using batchsize=128, accuracy =0.212

Conclusion:batchsize =8 performed best. however, when you look at the scores between the training and test set, batchsize = 128 has a smaller difference so we would preferably go with that as the first one has more overfitting

2. In this task, you'll implement several ANN models with different learning rates for the stochastic gradient descent. In all of the models below, use 128 as your mini batch size.

In [0]:
from tensorflow.keras import optimizers
sgd_001 = optimizers.SGD(lr=0.01)
sgd_100 = optimizers.SGD(lr=100)
sgd_00000001 = optimizers.SGD(lr=0.0000001)

2.1. Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using 0.01 as the learning rate.

In [56]:
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer=sgd_001, loss='categorical_hinge',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 34us/sample - loss: 1.0132 - accuracy: 0.3407
Epoch 2/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.9757 - accuracy: 0.5468
Epoch 3/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.8556 - accuracy: 0.6105
Epoch 4/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.6711 - accuracy: 0.6987
Epoch 5/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.5020 - accuracy: 0.8106
Epoch 6/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.3811 - accuracy: 0.8660
Epoch 7/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.3157 - accuracy: 0.8818
Epoch 8/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.2789 - accuracy: 0.8908
Epoch 9/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.2550 - accuracy: 0.8965
Epoch 10/20
60

In [57]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.1602322570681572
Test accuracy: 0.9311


2.2 Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using 100 as the learning rate.

In [58]:
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer=sgd_100, loss='categorical_hinge',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 35us/sample - loss: 1.7998 - accuracy: 0.0993
Epoch 2/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.8014 - accuracy: 0.0993
Epoch 3/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.8014 - accuracy: 0.0993
Epoch 4/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.8014 - accuracy: 0.0993
Epoch 5/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.8014 - accuracy: 0.0993
Epoch 6/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.8014 - accuracy: 0.0993
Epoch 7/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.8014 - accuracy: 0.0993
Epoch 8/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.8014 - accuracy: 0.0993
Epoch 9/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.8014 - accuracy: 0.0993
Epoch 10/20
60

In [59]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 1.7936
Test accuracy: 0.1032


2.3 Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using 0.0000001 as the learning rate.

In [60]:
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer=sgd_00000001, loss='categorical_hinge',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 35us/sample - loss: 1.0372 - accuracy: 0.0874
Epoch 2/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.0372 - accuracy: 0.0874
Epoch 3/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.0372 - accuracy: 0.0874
Epoch 4/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.0372 - accuracy: 0.0874
Epoch 5/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.0372 - accuracy: 0.0874
Epoch 6/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.0372 - accuracy: 0.0874
Epoch 7/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.0372 - accuracy: 0.0874
Epoch 8/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.0372 - accuracy: 0.0874
Epoch 9/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.0372 - accuracy: 0.0874
Epoch 10/20
60

In [61]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 1.0377731491088866
Test accuracy: 0.0816


RESULTS: Using 0.01, accuracy =0.931
      Using 100, accuracy =0.103
      Using 0.0000001, accuracy =0.0816

Conclusion: 0.01 seemed to perform best

In [0]:
#the model converges when the learning rate is 0.01. However, the model diverges when we used 100 as it is too large. When 0.0000001 is used, the model
#improves slowly. hence the 0.01 performs best